cli/util/uploader.go

```diff
@@ -0,0 +1,429 @@
+					}
+					offset := (uint(blockIndex) % indexFileSize) * oidSize
+					currOid.Encode(buffer[offset:])
+					oidFetcherToProcessor <- struct{}{}
+					return nil
+				})
+				if err != nil {
+					errCh <- err
+					return
+				}
+			}
+		}()
+	}
+
+	for i := existingIndexCount + 1; i <= expectedIndexCount; i++ {
+		startIndex := (i - 1) * indexFileSize
+		endIndex := startIndex + indexFileSize
+		go func() {
+			for j := int(startIndex); j < int(endIndex); j += searchBatchSize {
+				remaining := int(endIndex) - j
+				end := j + min(searchBatchSize, remaining)
+
+				prm := client.PrmObjectSearch{}
+				filters := object.NewSearchFilters()
+				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
+				filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
+				prm.SetFilters(filters)
+
+				objectIDs, err := neofs.ObjectSearch(ctx.Context, p, account.PrivateKey(), containerID.String(), prm)
+				if err != nil {
+					errCh <- fmt.Errorf("no OIDs found for index files: %w", err)
+					return
+				}
+
+				for _, currOid := range objectIDs {
+					oidCh <- currOid
+				}
+			}
+		}()
+
+		var completed int
+	waitLoop:
+		for {
+			select {
+			case err := <-errCh:
+				return err
+			case <-oidFetcherToProcessor:
+				completed++
+				if completed == int(indexFileSize) {
+					break waitLoop
+				}
+			}
+		}
+		select {
+		case err := <-errCh:
+			return err
+		default:
+		}
+		attrs := []object.Attribute{
+			*object.NewAttribute(attributeKey, strconv.Itoa(int(i))),
+			*object.NewAttribute("size", strconv.Itoa(int(indexFileSize))),
+		}
+		err = uploadObj(ctx.Context, *p, user.NewAutoIDSignerRFC6979(account.PrivateKey().PrivateKey), account.PrivateKey().GetScriptHash(), containerID, buffer, attrs, homomorphicHashingDisabled)
+		if err != nil {
+			return fmt.Errorf("failed to upload index file %d: %w", i, err)
+		}
+		fmt.Fprintf(ctx.App.Writer, "Uploaded index file %d\n", i)
+	}
+	return nil
+}
+
+// uploadObj uploads the block to the container using the pool.
+func uploadObj(ctx context.Context, p pool.Pool, signer user.Signer, owner util.Uint160, containerID cid.ID, objData []byte, attrs []object.Attribute, HomomorphicHashingDisabled bool) error {
+	var (
+		ownerID          user.ID
+		hdr              object.Object
+		ch               checksum.Checksum
+		v                = new(version.Version)
+		prmObjectPutInit client.PrmObjectPutInit
+	)
+
+	ownerID.SetScriptHash(owner)
+	hdr.SetPayload(objData)
+	hdr.SetPayloadSize(uint64(len(objData)))
+	hdr.SetContainerID(containerID)
+	hdr.SetOwnerID(&ownerID)
+	hdr.SetAttributes(attrs...)
+	hdr.SetCreationEpoch(1)
+	v.SetMajor(1)
+	hdr.SetVersion(v)
+	if !HomomorphicHashingDisabled {
+		checksum.Calculate(&ch, checksum.TZ, objData)
+		hdr.SetPayloadHomomorphicHash(ch)
+	}
+	hdr.SetPayloadHomomorphicHash(ch)
+	checksum.Calculate(&ch, checksum.SHA256, objData)
```

Comment:
And BTW, you're reusing `ch` between homomorphic and simple hash. Check that `ch` value is properly copied in `hdr.SetPayloadHomomorphicHash(ch)`